File: backend/intelligence/schema.py
from __future__ import annotations

from typing import Literal, Optional

from pydantic import BaseModel, Field


Moat = Literal["High", "Medium", "Low", "unknown"]
FinancialHealth = Literal["Pass", "Fail", "unknown"]
Valuation = Literal["Over-valued", "Under-valued", "unknown"]
Action = Literal["BUY", "HOLD", "SELL", "AVOID"]


class EvidenceItem(BaseModel):
    """
    Optional audit trail back to deterministic inputs.
    Keep this small; it's meant for debugging and portfolio rollups.
    """

    claim: str = Field(..., description="Short claim being supported")
    source: str = Field(..., description="Fact key or pointer, e.g. financials.forwardPE or latest_filing.snippets[2]")


class ReportModel(BaseModel):
    moat: Moat
    financial_health: FinancialHealth
    valuation: Valuation

    base_case: str
    upside_case: str
    downside_case: str

    action: Action
    reasoning: str

    evidence: Optional[list[EvidenceItem]] = None

File: frontend/src/intelligence/schema.ts
import { z } from "zod";

export const EvidenceItemSchema = z.object({
  claim: z.string(),
  source: z.string(),
});

export const ReportSchema = z.object({
  moat: z.enum(["High", "Medium", "Low", "unknown"]),
  financial_health: z.enum(["Pass", "Fail", "unknown"]),
  valuation: z.enum(["Over-valued", "Under-valued", "unknown"]),
  base_case: z.string(),
  upside_case: z.string(),
  downside_case: z.string(),
  action: z.enum(["BUY", "HOLD", "SELL", "AVOID"]),
  reasoning: z.string(),
  evidence: z.array(EvidenceItemSchema).optional().nullable(),
});

export type ReportFields = z.infer<typeof ReportSchema>;

export const ReportFileSchema = z.object({
  ticker: z.string(),
  asOfUtc: z.string(),
  provider: z.string().optional().nullable(),
  model: z.string().optional().nullable(),
  inputsMissing: z.array(z.string()).optional().nullable(),
  filingFetchStatus: z.string().optional().nullable(),
  latestFilingMeta: z.unknown().optional().nullable(),
  report: ReportSchema,
});

export type ReportFile = z.infer<typeof ReportFileSchema>;

